torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -1156,12 +1188,55 @@ std::vector<WorkerInfo> TensorPipeAgent::getWorkerInfos() const {
 
 const std::string& TensorPipeAgent::findWorkerURL(
     const WorkerInfo& worker) const {
-  const auto it = workerNameToURL_.find(worker.name_);
+  std::unordered_map<std::string, std::string>::const_iterator it;
+  if (isStaticGroup_) {
+    it = workerNameToURL_.find(worker.name_);
+  } else {
+    std::unique_lock<std::mutex> lock(groupMembershipMutex_);
+    it = workerNameToURL_.find(worker.name_);
+  }
```

Comment:
nit: since this code pattern has occurred multiple times, it might worth create a custom guard for it? E.g.

```c++
{
  SomeGuard guard(groupMembershipMutex_);
  it = workerNameToURL_.find(worker.name_);
}
```

Reply:
That is a great suggestion, thanks!